pkg/controller/volume/expand/expand_controller.go

```diff
@@ -0,0 +1,255 @@
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/volume/expand/cache"
+	"k8s.io/kubernetes/pkg/util/io"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/util/operationexecutor"
+)
+
+const (
+	syncLoopPeriod time.Duration = 100 * time.Millisecond
+)
+
+// ExpandController expands the pvs
+type ExpandController interface {
+	Run(stopCh <-chan struct{})
+}
+
+type expandController struct {
+	// kubeClient is the kube API client used by volumehost to communicate with
+	// the API server.
+	kubeClient clientset.Interface
+
+	// pvcLister is the shared PVC lister used to fetch and store PVC
+	// objects from the API server. It is shared with other controllers and
+	// therefore the PVC objects in its store should be treated as immutable.
+	pvcLister  corelisters.PersistentVolumeClaimLister
+	pvcsSynced kcache.InformerSynced
+
+	pvLister corelisters.PersistentVolumeLister
+	pvSynced kcache.InformerSynced
+
+	// cloud provider used by volume host
+	cloud cloudprovider.Interface
+
+	// volumePluginMgr used to initialize and fetch volume plugins
+	volumePluginMgr volume.VolumePluginMgr
+
+	// recorder is used to record events in the API server
+	recorder record.EventRecorder
+
+	// Volume resize map of volumes that needs resizing
+	resizeMap cache.VolumeResizeMap
+
+	// Worker goroutine to process resize requests from resizeMap
+	syncResize SyncVolumeResize
+
+	// Operation executor
+	opExecutor operationexecutor.OperationExecutor
+}
+
+func NewExpandController(
+	kubeClient clientset.Interface,
+	pvcInformer coreinformers.PersistentVolumeClaimInformer,
+	pvInformer coreinformers.PersistentVolumeInformer,
+	cloud cloudprovider.Interface,
+	plugins []volume.VolumePlugin) (ExpandController, error) {
+
+	expc := &expandController{
+		kubeClient: kubeClient,
+		cloud:      cloud,
+		pvcLister:  pvcInformer.Lister(),
+		pvcsSynced: pvcInformer.Informer().HasSynced,
+		pvLister:   pvInformer.Lister(),
+		pvSynced:   pvInformer.Informer().HasSynced,
+	}
+
+	if err := expc.volumePluginMgr.InitPlugins(plugins, nil, expc); err != nil {
+		return nil, fmt.Errorf("Could not initialize volume plugins for Expand Controller : %+v", err)
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "volume_expand"})
+
+	expc.opExecutor = operationexecutor.NewOperationExecutor(operationexecutor.NewOperationGenerator(
+		kubeClient,
+		&expc.volumePluginMgr,
+		recorder,
+		false))
+
+	expc.resizeMap = cache.NewVolumeResizeMap(expc.kubeClient)
+
+	pvcInformer.Informer().AddEventHandler(kcache.ResourceEventHandlerFuncs{
+		UpdateFunc: expc.pvcUpdate,
```

Comment:
Watching PVC update is not enough. You need to react to PVCs that were modified when the controller was down. I.e. AddFunc + check that PVC.Spec.Capacity > PVC.Status.Capacity and then check the bound PV.Spec.Capacity.

Reply:
Reaction to Delete event might be useful too, as `resizeMap` should be updated when user deletes the PVC and thus is not interested in its resize.

Reply:
We will add a separate populator loop to handle pvc updates while controller was down (not in this PR)

Reply:
I still miss reaction to Added events. When the controller is down when user resizes a PVC, all you get is a PVC Added event with different pvc.status.capacity and spec.capacity. You need to react to this.

Which brings me back to periodical reconcilation of all PVCs and just checking their pvc.status.capacity vs. pvc.spec.capacity. You don't need caches / queues and it's bulletproff to missed events (e.g. when the controller is down).

Reply:
Are you saying, we should replace event based handling of PVCs with periodically reconcilation of all PVCs? so basically - replace current event based design with one based on poll. Will that be *only* way of handling resizes or we still watch for events?

Reply:
I am not opposed to periodic reconcilation of all PVCs via polling with some sort of populator loop. I will add that in a follow up PR. It is mentioned in design doc too. Such a populator loop will run less frequently though, to avoid causing high cpu usage. 

@jsafrane you seem to be implying that, if we used periodic polling of all PVCs as *only* way of finding out PVCs which needed resizing then we won't need to cache/queue the PVCs.  While that is correct, in a large cluster with 10K PVCs or more, we will causing quite a bit of CPU usage. On other hand -  if we keep polling interval too long, then there could be significant delay in time user submits resize request and when resize requests get processed.

Reply:
I think he means two things:
- we need `AddFunc` to handle Add events, not just `UpdateFunc`
- we should do periodic reconciliation of *all* PVCs to handle missed events. there is no need for a complicated cache in this case: AddFunc + UpdateFunc + periodic reconciliation will take care of everything. i.e. instead of having AddFunc/UpdateFunc add to a cache which is then acted upon, simply act as soon as AddFunc/UpdateFunc occurs.

Reply:
Added a separate pvc_populator loop to handle case of missed events because of controller crash.